zb-db/src/main/java/io/zeebe/db/impl/rocksdb/ZeebeRocksDbFactory.java

```diff
@@ -55,84 +66,196 @@ private ZeebeRocksDbFactory(
+    // this is a current guess and candidate for further tuning
+    // values can be between 0 and 0.25 (anything higher gets clamped to 0.25), we randomly picked
+    // 0.15
+    // prefix seek must be fast, so we allocate some extra memory of a single memtable budget to
+    // create
+    // a filter for each memtable, allowing us to skip the prefixes if possible
+    final var memtablePrefixFilterMemory = 0.15;
+    final var memtableMemory =
+        Math.round(
+            ((totalMemoryBudget - blockCacheMemory) / (double) maxConcurrentMemtableCount)
+                * (1 - memtablePrefixFilterMemory));
+
+    final var tableConfig = createTableFormatConfig(closeables, blockCacheMemory);
+
+    return columnFamilyOptions
+        // to extract our column family type (used as prefix) and seek faster
+        .useFixedLengthPrefixExtractor(Long.BYTES)
+        .setMemtablePrefixBloomSizeRatio(memtablePrefixFilterMemory)
+        // memtables
+        // merge at least 3 memtables per L0 file, otherwise all memtables are flushed as individual
+        // files
+        // this is also a candidate for tuning, it was a rough guess
+        .setMinWriteBufferNumberToMerge(Math.min(3, maxConcurrentMemtableCount))
+        .setMaxWriteBufferNumberToMaintain(maxConcurrentMemtableCount)
+        .setMaxWriteBufferNumber(maxConcurrentMemtableCount)
+        .setWriteBufferSize(memtableMemory)
+        // compaction
+        .setLevelCompactionDynamicLevelBytes(true)
+        .setCompactionPriority(CompactionPriority.OldestSmallestSeqFirst)
+        .setCompactionStyle(CompactionStyle.LEVEL)
+        // L-0 means immediately flushed memtables
+        .setLevel0FileNumCompactionTrigger(maxConcurrentMemtableCount)
+        .setLevel0SlowdownWritesTrigger(
+            maxConcurrentMemtableCount + (maxConcurrentMemtableCount / 2))
+        .setLevel0StopWritesTrigger(maxConcurrentMemtableCount * 2)
+        // configure 4 levels: L1 = 32mb, L2 = 320mb, L3 = 3.2Gb, L4 >= 3.2Gb
+        // level 1 and 2 are uncompressed, level 3 and above are compressed using a CPU-cheap
+        // compression algo. compressed blocks are stored in the OS page cache, and uncompressed in
+        // the LRUCache created above. note L0 is always uncompressed
+        .setNumLevels(4)
+        .setMaxBytesForLevelBase(32 * 1024 * 1024L)
+        .setMaxBytesForLevelMultiplier(10)
+        .setCompressionPerLevel(
+            List.of(
+                CompressionType.NO_COMPRESSION,
+                CompressionType.NO_COMPRESSION,
+                CompressionType.LZ4_COMPRESSION,
+                CompressionType.LZ4_COMPRESSION))
+        // defines the desired SST file size (but not guaranteed, it is usually lower)
+        // L0 => 8Mb, L1 => 16Mb, L2 => 32Mb, L3 => 64Mb
+        // as levels get bigger, we want to have a good balance between the number of files and the
+        // individual file sizes
+        .setTargetFileSizeBase(8 * 1024 * 1024L)
```

Comment:
`minWriteBufferToMerge` just describes the minimum it should wait until it can flush - not the exact number it will be, so it can always be greater. From experience, I could never get the file size to be within expected range without the targetFileSize setting.

Reply:
```
  // Target file size for compaction.
  // target_file_size_base is per-file size for level-1.
  // Target file size for level L can be calculated by
  // target_file_size_base * (target_file_size_multiplier ^ (L-1))
  // For example, if target_file_size_base is 2MB and
  // target_file_size_multiplier is 10, then each file on level-1 will
  // be 2MB, and each file on level 2 will be 20MB,
  // and each file on level-3 will be 200MB.
  //
  // Default: 64MB.
  //
  // Dynamically changeable through SetOptions() API
  uint64_t target_file_size_base = 64 * 1048576;
```

https://github.com/facebook/rocksdb/blob/fd0d35d390e212b617e90d7567102d3e5fd1c706/include/rocksdb/advanced_options.h#L417-L429

Reply:
So my comment is wrong then, L0 is what gets flushed, e.g. 3 to X memtables, and target file size is for L1. So L1 is 8mb, L2 is 16MB, L3 is 32MB only. Or?

Reply:
Yes correct.

Reply:
i will update the comment then